### PR TITLE
added rest xds server for endpoints

### PR DESCRIPTION
**Pull request by EItanya**
# Description

Allow toggling of EDS to rest XDS to avoid the envoy issue where updates to CDS require new endpoints from the control-plane

# Context
https://github.com/envoyproxy/envoy/issues/13070.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/3673
resolves https://github.com/solo-io/gloo/issues/3710
resolves https://github.com/solo-io/gloo/issues/3219
resolves https://github.com/solo-io/gloo/issues/3805

### CODE REVIEW

**Review by kdorosh — CHANGES_REQUESTED** [review]
please updated the pr description

File: changelog/v1.6.0-beta6/rest-eds.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    description: Fix EDS so modifying configs on a TLS enabled Upstream no longer results in 503s
+    resolvesIssue: true
+  - type: FIX
+    issueLink: https://github.com/solo-io/gloo/issues/3710
+    description: Gloo will no longer remove upstreams after editing their spec.
+    resolvesIssue: true
+  - type: FIX
+    issueLink: https://github.com/solo-io/gloo/issues/3673
+    description: Fix EDS so modifying health checks on Upstream no longer results in 503s
+    resolvesIssue: true
```
Comment (kdorosh) [inline changelog/v1.6.0-beta6/rest-eds.yaml:13]: this is duplicated, remove

File: install/test/helm_suite_test.go [hunk 2]
```diff
@@ -300,8 +302,19 @@ func mergeMaps(a, b map[string]interface{}) map[string]interface{} {
...
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
 	runtimeObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, jsn)
-	Expect(err).NotTo(HaveOccurred())
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
 	return runtimeObj.(*unstructured.Unstructured)
 }
+
+func makeUnstructureFromTemplateFile(fixtureName string, values interface{}) *unstructured.Unstructured {
+	// file, err := ioutil.ReadFile(fixtureName)
+	// ExpectWithOffset(1, err).NotTo(HaveOccurred())
```
Comment (kdorosh) [inline install/test/helm_suite_test.go:313]: remove

File: projects/gloo/api/v1/settings.proto [hunk 3]
```diff
@@ -440,9 +440,18 @@ message GlooOptions {
...
+    // Where the `gloo` REST xDS server should bind.
     // Defaults to `0.0.0.0:9976`
     string rest_xds_bind_addr = 11;
+
+    // Whether or not to use rest xds for all EDS by default
+    // This setting is meant to solve the bug which causes updated upstreams to dissapear, or have 0 endpoints. 
+    // Some examples are:
+    // 1. https://github.com/solo-io/gloo/issues/3673
+    // 2. https://github.com/solo-io/gloo/issues/3710
+    // 3. https://github.com/solo-io/gloo/issues/3673
```
Comment (kdorosh) [inline projects/gloo/api/v1/settings.proto:452]: this is duplicated. perhaps link to the envoy issue instead?

File: projects/gloo/api/v1/settings.proto [hunk 4]
```diff
@@ -440,9 +440,18 @@ message GlooOptions {
...
+    // Rest XDS, as opposed to grpc, uses http polling rather than streaming
+    google.protobuf.BoolValue enable_rest_eds = 12;
```
Comment (kdorosh) [inline projects/gloo/api/v1/settings.proto:454]: what is the default?

File: changelog/v1.6.0-beta6/rest-eds.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  - type: NEW_FEATURE
+    issueLink: https://github.com/solo-io/gloo/issues/3805
+    description: >-
+      Allow toggling of EDS to rest XDS to avoid the envoy issue described in the following issue: https://github.com/envoyproxy/envoy/issues/13070.
```
Comment (kdorosh) [inline changelog/v1.6.0-beta6/rest-eds.yaml:17]: what is the default? this was a change from the original behavior before, correct?

File: test/e2e/happypath_test.go [hunk 6]
```diff
@@ -52,6 +53,24 @@ var _ = Describe("Happy path", func() {
...
+					Value: true,
+				},
+			},
+			{
+				Title: "Rest Eds Disabled",
+				RestEdsEnabled: &types.BoolValue{
+					Value: false,
+				},
+			},
+		}
```
Comment (kdorosh) [inline test/e2e/happypath_test.go:73]: this feels like it should be a `Table` with `Entry`s

Comment (kdorosh) [reply test/e2e/happypath_test.go:73]: i see, you use it to wrap a bunch of other tests 👍

File: install/helm/gloo/templates/18-settings.yaml [hunk 7]
```diff
@@ -15,6 +15,7 @@ spec:
 {{- else }}
     xdsBindAddr: "0.0.0.0:{{ .Values.gloo.deployment.xdsPort }}"
     restXdsBindAddr: "0.0.0.0:{{ .Values.gloo.deployment.restXdsPort }}"
+    enableRestEds: true
```
Comment (kdorosh) [inline install/helm/gloo/templates/18-settings.yaml:18]: we should have a helm value for this

File: projects/gloo/api/v1/settings.proto [hunk 8]
```diff
@@ -440,9 +440,19 @@ message GlooOptions {
     // defaults to 100.
     google.protobuf.UInt32Value regex_max_program_size = 10;
 
-    // (Enterprise Only): Where the `gloo` REST xDS server should bind. Used by Gloo Federation.
+    // Where the `gloo` REST xDS server should bind.
     // Defaults to `0.0.0.0:9976`
     string rest_xds_bind_addr = 11;
+
+    // Whether or not to use rest xds for all EDS by default. 
+    // Set to true by default in versions > `v1.6.0`.
```
Comment (kdorosh) [inline projects/gloo/api/v1/settings.proto:448]: only set to true in helm installations, we should fix the implementation to handle nil as true

**Review by kdorosh — APPROVED** [review]

**Review by kdorosh — APPROVED** [review]

File: go.mod [hunk 9]
```diff
@@ -66,7 +66,7 @@ require (
 	github.com/solo-io/reporting-client v0.1.2
 	github.com/solo-io/skv2 v0.8.1
 	github.com/solo-io/solo-apis v0.0.0-20200717214114-6a1daa5a5d05
-	github.com/solo-io/solo-kit v0.13.13
+	github.com/solo-io/solo-kit v0.13.14
```
Comment (kdorosh) [inline go.mod:69]: nothing in the changelog for the codegen related changes from this?

File: install/helm/gloo/templates/18-settings.yaml [hunk 10]
```diff
@@ -15,6 +15,7 @@ spec:
 {{- else }}
     xdsBindAddr: "0.0.0.0:{{ .Values.gloo.deployment.xdsPort }}"
     restXdsBindAddr: "0.0.0.0:{{ .Values.gloo.deployment.restXdsPort }}"
+    enableRestEds: {{ .Values.settings.enableRestEds | default true }}
```
Comment (kdorosh) [inline install/helm/gloo/templates/18-settings.yaml:18]: i don't understand why we need the default -- shouldn't the generated values.go file already have the true value?

File: go.mod (hunk 9 above)
Comment (EItanya) [reply go.mod:69]: I can add a DEPENDENCY_BUMP. This fixed some errors in the rest xds server

**Review by kdorosh — APPROVED** [review]

**Review by kdorosh — APPROVED** [review]

**Pull request merged**